Distinguish (internally) between legacy and new passphrase encryptions

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#1436 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- c433d811172cf0983ceb741a304ea9785958e7f5
  Augment passphrase metadata info with encryptiohttps://github.com/input-output-hk/cardano-wallet/compare/paweljakubas/1436/byron-wallet-restoration-from-xprv-endpoint-impl...KtorZ/1436/encrypt-with-scrypt?expand=1n scheme
  This to allow distinguish legacy passphrase hashes imported from Daedalus that are
encrypted through Scrypt and the new passphrases that we now encrypt with PBKDF2.

- 413498eb1aa0d389ec2877465d243daaa76743a7
  Actually implement 'checkPassphrase' and 'encryptPassphrase' for Scrypt
  This still misses some test vectors to compare with the legacy implementation. I'll
produce that soon using both Daedalus + cardano-sl code.

# Comments

<!-- Additional comments or screenshots to attach if any -->



<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
Augment passphrase metadata info with encryption scheme

This to allow distinguish legacy passphrase hashes imported from Daedalus that are
encrypted through Scrypt and the new passphrases that we now encrypt with PBKDF2.

KtorZ committed:
Actually implement 'checkPassphrase' and 'encryptPassphrase' for Scrypt

This still misses some test vectors to compare with the legacy implementation. I'll
produce that soon using both Daedalus + cardano-sl code.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -57,10 +57,11 @@ share
+    walCreationTime             UTCTime                   sql=creation_time
+    walName                     Text                      sql=name
+    walPassphraseLastUpdatedAt  UTCTime Maybe             sql=passphrase_last_updated_at
+    walPassphraseScheme         W.PassphraseScheme Maybe  sql=passphrase_scheme
```
I bet this will require some database migration, I need to look into this.

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -495,27 +499,60 @@ instance
+-- Yes there's CBOR. Yes it's hashed before being encrypted. You don't actually
+-- pronounce the 'l' in salmon. A lot of things in the world don't make sense
+-- but we still live just fine.
+preparePassphrase
```
probably we will need to export it in the module as we will need it in integration test or even better to expose : 
```
        salt <- Scrypt.Salt <$> getRandomBytes @_ @ByteString 32
        let msg = preparePassphrase pwd
        return $ Hash $ Scrypt.getEncryptedPass $
            Scrypt.encryptPass Scrypt.defaultParams salt msg
```
in a function to avoid importing Scrypt and Cbor in core-integration

KtorZ replied:
We already expose `encryptPassphrase` and `checkPassphrases`, and that's all which is needed :)

paweljakubas replied:
indeed!

paweljakubas reviewed (approved):
very good idea to introduce `PassphraseScheme` 
It would be good also to : 
1. update swagger's https://github.com/input-output-hk/cardano-wallet/blob/master/specifications/api/swagger.yaml#L279 if it is not 128 bytes
2. finish integration test showing restoration for byron wallet from master root key works

we will need also to have patch nix stuff.

KtorZ committed:
rework the checkPassphrase / encryptPassphrase phantom types

Wallet using a passphrase that is encrypted with the legacy scheme needs to go through some special
transformation before being encrypted. In order to present programmer mistakes here, we leverage a bit
the phantom types for 'Passphrase' such that passphrases obtained from the API are marked as 'raw' whereas,
every function from the 'AddressDerivation' module requires a passphrase to be marked as 'encryption'.
Such passphrases can be obtained via 'preparePassphrase' which does nothing but a simple type coercion,
for legacy passphrases it'll do some preliminary transformation.

KtorZ committed:
add golden test about legacy passphrase encryption

The golden string was obtained from cardano-sl, using a raw string that was blake2b_256-encoded by
Daedalus.
